torch/csrc/cuda/nccl.cpp

```diff
@@ -35,30 +38,35 @@ struct NcclCommList {
            In these cases, skip ncclCommDestroy */
           return;
         }
-	ncclCommDestroy(comms[i]);
+        ncclCommDestroy(comms[i]);
       }
     }
   }
+  ArrayRef<ncclComm_t> ref() const {
+    return ArrayRef<ncclComm_t>(comms.get(), ndevices);
+  }
 };
 
 // accesses to this object have to be guarded by THC's CudaFreeMutex
-std::unordered_map<std::string, NcclCommList > _communicators;
+std::unordered_map<std::string, NcclCommList> _communicators;
 
-static ncclComm_t* _get_communicator(std::vector<at::Tensor>& inputs) {
-  int ndevices = inputs.size();
+static ArrayRef<ncclComm_t> _get_communicators(TensorList inputs, ArrayRef<ncclComm_t> user_comms) {
+  if (user_comms.size() == inputs.size()) {
+    return user_comms;
```

Comment:
I'd say that doing no checks for `user_comms` and silently falling back to regular operation if they don't match the inputs is a bit weird. You give the comms, but you can't even be sure if they are used. This should probably be an error.